crates/dojo-erc/src/erc1155/erc1155.cairo

```diff
@@ -0,0 +1,449 @@
+            operator: ContractAddress,
+            from: ContractAddress,
+            token_ids: Array<u256>,
+            amounts: Array<u256>,
+            data: Array<u8>
+        ) -> u32;
+    }
+
+    #[starknet::interface]
+    trait IERC165 {
+        fn supports_interface(self: ContractState, interface_id: u32) -> bool;
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        TransferSingle: TransferSingle,
+        TransferBatch: TransferBatch,
+        ApprovalForAll: ApprovalForAll
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct TransferSingle {
+        operator: ContractAddress,
+        from: ContractAddress,
+        to: ContractAddress,
+        id: u256,
+        value: u256
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct TransferBatch {
+        operator: ContractAddress,
+        from: ContractAddress,
+        to: ContractAddress,
+        ids: Array<u256>,
+        values: Array<u256>
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct ApprovalForAll {
+        account: ContractAddress,
+        operator: ContractAddress,
+        approved: bool
+    }
+
+    #[storage]
+    struct Storage {
+        world_address: ContractAddress,
+    }
+
+    //
+    // Constructor
+    //
+
+    #[constructor]
+    fn constructor(ref self: ContractState, world: ContractAddress, uri: felt252) {
+        self.world_address.write(world);
+        _set_uri(ref self, uri);
+    }
+
+    #[view]
+    fn supports_interface(interface_id: u32) -> bool {
+        interface_id == INTERFACE_ERC165 ||
+        interface_id == INTERFACE_ERC1155 ||
+        interface_id == INTERFACE_ERC1155_METADATA
+    }
+
+    //
+    // ERC1155Metadata
+    //
+
+    #[view]
+    fn uri(ref self: ContractState, token_id: u256) -> felt252 {
+        let token = get_contract_address();
+        let query: Query = (token, (u256_as_allowance(token_id))).into_partitioned();
+        let mut uri_raw = world(@self).entity('Uri'.into(), query, 0, 0);
+        let _uri = serde::Serde::<Uri>::deserialize(ref uri_raw).unwrap();
+        _uri.uri
+    }
+
+    //
+    // ERC1155
+    //
+
+    #[view]
+    fn balance_of(ref self: ContractState, account: ContractAddress, id: u256) -> u256 {
+        let token = get_contract_address();
+        let query: Query = (token, (account, u256_as_allowance(id))).into_partitioned();
+        let mut balance_raw = world(@self).entity('Balance'.into(), query, 0, 0);     
+        let balance = serde::Serde::<Balance>::deserialize(ref balance_raw).unwrap();
+        balance.amount.into()
+    }
+
+    #[view]
+    fn balance_of_batch(ref self: ContractState, accounts: Array<ContractAddress>, ids: Array<u256>) -> Array<u256> {
+        assert(ids.len() == accounts.len(), 'ERC1155: invalid length');
+
+        let mut batch_balances = ArrayTrait::new();
+        let mut index = 0;
+        loop {
+            if index == ids.len() {
+                break();
```

Comment:
You can also `break batch_balances;` here and have the `loop {}` be the last statement in the fn so that it returns. It's cleaner IMO.